### DISCUSSION TITLE
How does `expanded` state work?

### DISCUSSION THREAD

**Discussion by githorse** [TanStack/table#4862] [Q&A]
[`ExpandedState`](https://tanstack.com/table/v8/docs/api/features/expanding) looks like this:

```
export type ExpandedState = true | Record<string, boolean>
```

From this I expected the `string` to be some kind of unique ID on the row -- but best I can tell from the [example](https://codesandbox.io/s/bold-nova-42k3vg?file=/src/main.tsx), it's actually a map of row index (i.e., a number) to `boolean`.

Questions:
1. Shouldn't this be a `Record<number, boolean>`? Or is it not always the row index?
3. What happens to expanded rows if I sort or filter my records? Do I need to manually update the `expanded` state?
4. Would it be preferable to keep the state as a `Record<unique row id, boolean>` so that the state persists across sorting, filtering, etc? (I don't know if it's possible to define a unique id on a row with Tanstack; many grids (e.g. MUI's) require that every record have a unique ID for this purpose.)
5. Could the docs be expanded with more detail on the *meaning* of the expanded state, rather than just the types?

On the final point -- in general the Tanstack docs seem to do a great job of telling me the Typescript types I'm dealing with -- much appreciated -- but not such a hot job of telling me what anything actually means. The [Expanded State API section](https://tanstack.com/table/v8/docs/api/features/expanding#state) would be far more useful if it included a simple comment like "expanded state is a map from the row index to a boolean" (or whatever it actually is). As it stands I'm left trying to piece this together from the example. (Obviously it would be even *more* helpful with notes about what happens on sorting & filtering, etc.)

**Comment by MorningLit** [comment]
Same here, I'm struggling to understand this library, it doesn't make sense that we have to manually update the expanded rows state when sorting / filtering, this should be handled by this library
What's the point of giving us full control when we have to spend hours trying to understand how this library is supposed to be used?
And the fact that this library did a terrible job of explaining the API does not help either

**Reply by TonyVeigel** [reply]
Why would you need to update the expanded state when sorting or filtering? the expanded state takes in a row number so if you're sorting it is attached to the row number, same with filtering. Or what you trying to accomplish?

**Comment by githorse** [comment]
I figured out why this is `string`, not `number`: because you can define a unique row ID with [getRowId](https://tanstack.com/table/v8/docs/api/core/table#getrowid), and the `ExpandedState` will use that ID instead of the index. This also applies to [row selection](https://tanstack.com/table/v8/docs/api/features/row-selection) and probably elsewhere as well.

This answers my questions (1) - (3), above:

1. It's not always the row index; if you define `getRowId`, it will be a unique ID. `string` covers all the cases.
2. Using a unique ID instead of the index resolves issues with sorting or filtering changing the indices.
3. Yep. It's better to define `getRowId`.

This all seems pretty sensible (though I would argue that the type should just be `string | number` and it should actually use numbers for indices, not strings.) But still no joy on point (4) -- updating the docs to explain this.

**Reply by pere-seapoint** [reply]
You saved the day. Thank you!

**Reply by pere-seapoint** [reply]
And to answer your question 4: yes, absolutely, its is confusing as hell
